File: lib/assets/js/hooks/toast.js
/**
 * PureAdminToast - Toast notification system.
 *
 * Attach to a toast container element. The server triggers toasts via push_event,
 * and the client handles rendering, auto-dismiss, and close button — no server
 * round-trips for display/dismiss.
 *
 * Server usage (in LiveView):
 *
 *   push_toast(socket, "success", "Saved!", "Changes saved.")
 *   push_toast(socket, "danger", "Error", "Failed.", duration: 0)
 *   push_toast(socket, "info", "Note", "FYI", progress: true, filled: true)
 *   push_toast(socket, "warning", "Confirm", "Delete?",
 *     actions: [
 *       %{label: "Undo", event: "undo", params: %{id: 1}, variant: "warning"},
 *       %{label: "Dismiss", dismiss: true}
 *     ])
 *
 * Template:
 *
 *   <.toast_container id="toasts" position="top-end" is_hook />
 *
 * Multiple containers can exist for different positions. Each container only
 * shows toasts matching its position (from data-position attr).
 */
export const PureAdminToast = {
  mounted() {
    this._toastId = 0
    this._position = this.el.dataset.position || 'top-end'

    this.handleEvent("toast", (toast) => {
      const position = toast.position || 'top-end'
      if (position !== this._position) return

      this._showToast(toast)
    })

    // Also support adding toasts from other hooks/JS via custom event
    this.el.addEventListener("pa:toast", (e) => {
      this._showToast(e.detail)
    })
  },

  _showToast(toast) {
    const id = `pa-toast-${this._position}-${this._toastId++}`
    const variant = toast.variant || 'info'
    const title = toast.title || ''
    const message = toast.message || ''
    const duration = toast.duration !== undefined ? toast.duration : 5000
    const filled = toast.filled || false
    const progress = toast.progress || false
    const progressColor = toast.progress_color || null
    const maxWidth = toast.max_width || null
    const actions = toast.actions || []

    const variantClass = filled ? `pa-toast--filled-${variant}` : `pa-toast--${variant}`
    const el = document.createElement('div')
    el.id = id
    el.className = `pa-toast ${variantClass} pa-toast--show`
    if (maxWidth) el.style.maxWidth = maxWidth

    // Build actions HTML
    let actionsHtml = ''
    if (actions.length > 0) {
      const btns = actions.map(a =>
        `<button class="pa-btn pa-btn--xs pa-btn--${a.variant || 'secondary'}" data-action="${this._escapeAttr(JSON.stringify(a))}">${this._escapeHtml(a.label || 'Action')}</button>`
      ).join('')
      actionsHtml = `<div class="pa-toast__actions">${btns}</div>`
    }

    // Build progress bar HTML
    const progressStyle = progressColor
      ? `width: 100%; color: ${progressColor};`
      : 'width: 100%;'
    const progressHtml = progress && duration > 0
      ? `<div class="pa-toast__progress" style="${progressStyle}"></div>`
      : ''

    el.innerHTML = `
      <div class="pa-toast__content">
        ${title ? `<div class="pa-toast__title">${this._escapeHtml(title)}</div>` : ''}
        ${message ? `<div class="pa-toast__message">${this._escapeHtml(message)}</div>` : ''}
        ${actionsHtml}
      </div>
      <button class="pa-toast__close" aria-label="Close">
        <svg xmlns="http://www.w3.org/2000/svg" width="18" height="18" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><path d="M18 6 6 18"/><path d="m6 6 12 12"/></svg>
      </button>
      ${progressHtml}
    `

    // Close button
    el.querySelector('.pa-toast__close').addEventListener('click', () => {
      this._dismissToast(el)
    })

    // Action button handlers
    if (actions.length > 0) {
      el.querySelectorAll('.pa-toast__actions [data-action]').forEach((btn) => {
        btn.addEventListener('click', (e) => {
          e.stopPropagation()
          const action = JSON.parse(btn.dataset.action)
          if (action.event) {
            this.pushEvent(action.event, action.params || {})
          }
          this._dismissToast(el)
        })
      })
    } else {
      // Toasts without actions are click-to-dismiss
      el.style.cursor = 'pointer'
      el.addEventListener('click', () => this._dismissToast(el))
    }

    this.el.appendChild(el)

    // Ratchet: lock container to its peak width
    requestAnimationFrame(() => {
      const currentWidth = this.el.offsetWidth
      const peakWidth = parseInt(this.el.dataset.peakWidth || '0', 10)
      if (currentWidth > peakWidth) {
        this.el.dataset.peakWidth = currentWidth
        this.el.style.minWidth = currentWidth + 'px'
      }
    })

    // Progress bar animation
    if (progress && duration > 0) {
      const progressEl = el.querySelector('.pa-toast__progress')
      if (progressEl) {
        progressEl.style.transition = `width ${duration}ms linear`
        setTimeout(() => { progressEl.style.width = '0%' }, 50)
      }
    }

    // Auto-dismiss
    if (duration > 0) {
      setTimeout(() => this._dismissToast(el), duration)
    }
  },

  _dismissToast(el) {
    if (!el || !el.parentNode) return
    el.classList.remove('pa-toast--show')
    el.classList.add('pa-toast--dismissing')
    setTimeout(() => {
      if (el.parentNode) {
        const container = el.parentNode
        container.removeChild(el)

        // Reset peak width when container is empty
        if (!container.querySelector('.pa-toast')) {
          container.style.minWidth = ''
          container.dataset.peakWidth = '0'
        }
      }
    }, 300)
  },

  _escapeHtml(text) {
    const div = document.createElement('div')
    div.textContent = text
    return div.innerHTML
  },

  _escapeAttr(text) {
    return text.replace(/&/g, '&amp;').replace(/"/g, '&quot;')
  }
}

File: lib/keen_pure_admin/components/toast.ex
defmodule PureAdmin.Components.Toast do
  @moduledoc """
  Toast notification components for Pure Admin.

  Provides `toast/1` for server-rendered toasts, `toast_container/1` for positioning,
  and `push_toast/3` for triggering client-side toasts via `push_event`.

  ## Client-side toasts (recommended)

  Add a hook-enabled container in your layout:

      <.toast_container id="toasts" position="top-end" is_hook />

  Then push toasts from any LiveView:

      socket |> push_toast("success", "Saved!", "Your changes have been saved.")

  The JS hook handles rendering and auto-dismiss — no server round-trips.
  """
  use Phoenix.Component

  import PureAdmin.Helpers

  @doc """
  Pushes a toast notification to the client via `push_event`.

  The client-side `PureAdminToast` hook renders and auto-dismisses the toast.

  ## Options

  - `:duration` — auto-dismiss in ms (default: 5000, 0 = persistent)
  - `:position` — override container position (default: "top-end")

  ## Examples

      socket |> push_toast("success", "Saved!", "Changes saved successfully.")
      socket |> push_toast("danger", "Error", "Something went wrong.", duration: 0)
      socket |> push_toast("info", "Note", "FYI", position: "bottom-end")
  """
  def push_toast(socket, variant, title, message, opts \\ []) do
    payload = %{
      variant: variant,
      title: title,
      message: message,
      duration: Keyword.get(opts, :duration, 5000),
      position: Keyword.get(opts, :position, "top-end"),
      filled: Keyword.get(opts, :filled, false),
      progress: Keyword.get(opts, :progress, false)
    }

    payload =
      payload
      |> maybe_put(:progress_color, Keyword.get(opts, :progress_color))
      |> maybe_put(:max_width, Keyword.get(opts, :max_width))
      |> maybe_put(:actions, Keyword.get(opts, :actions))

    Phoenix.LiveView.push_event(socket, "toast", payload)
  end

  defp maybe_put(map, _key, nil), do: map
  defp maybe_put(map, key, value), do: Map.put(map, key, value)

  @doc """
  Renders a toast notification.

  ## Examples

      <.toast variant="success" title_text="Success!" message_text="Changes saved." />

      <.toast variant="danger" title_text="Error" message_text="Save failed." on_close="dismiss_toast" />
  """
  attr(:id, :string, default: nil)
  attr(:variant, :string, default: "info",
    values: ["primary", "success", "danger", "warning", "info"])
  attr(:theme_color, :string, default: nil,
    values: [nil, "1", "2", "3", "4", "5", "6", "7", "8", "9"],
    doc: "Theme color slot 1-9 (overrides variant)")
  attr(:is_filled, :boolean, default: false,
    doc: "Filled style with full-color background and contrast text")
  attr(:title_text, :string, default: nil, doc: "Toast title")
  attr(:message_text, :string, default: nil, doc: "Toast message")
  attr(:is_visible, :boolean, default: true, doc: "Show/hide the toast")
  attr(:on_close, :string, default: nil, doc: "LiveView event fired on close")
  attr(:class, :string, default: nil)
  attr(:rest, :global, include: ~w(phx-click phx-value-id))
  slot(:icon, doc: "Custom icon content")
  slot(:inner_block, doc: "Custom body content (overrides title_text/message_text)")

  def toast(assigns) do
    ~H"""
    <div
      :if={@is_visible}
      id={@id}
      class={build_classes("pa-toast", [
        {toast_variant_class(assigns), true},
        {"pa-toast--show", @is_visible}
      ], @class)}
      {@rest}
    >
      <div :if={@icon != []} class="pa-toast__icon">
        <%= for icon <- @icon do %>
          <%= render_slot(icon) %>
        <% end %>
      </div>
      <%= if @inner_block != [] do %>
        <%= render_slot(@inner_block) %>
      <% else %>
        <div class="pa-toast__content">
          <div :if={@title_text} class="pa-toast__title"><%= @title_text %></div>
          <div :if={@message_text} class="pa-toast__message"><%= @message_text %></div>
        </div>
      <% end %>
      <button
        :if={@on_close}
        class="pa-toast__close"
        phx-click={@on_close}
        phx-value-id={@id}
        aria-label="Close"
      >
        <svg xmlns="http://www.w3.org/2000/svg" width="18" height="18" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><path d="M18 6 6 18"/><path d="m6 6 12 12"/></svg>
      </button>
    </div>
    """
  end

  @doc """
  Renders a toast container for positioning toasts in the viewport.

  ## Positions

  `top-end`, `top-center`, `top-start`, `bottom-end`, `bottom-center`, `bottom-start`

  ## Examples

      <.toast_container position="top-end">
        <.toast :for={t <- @toasts} id={t.id} variant={t.variant} title_text={t.title} message_text={t.message} on_close="dismiss_toast" />
      </.toast_container>
  """
  attr(:id, :string, default: nil, doc: "Required when using phx-hook")
  attr(:position, :string, default: "top-end",
    values: ["top-end", "top-center", "top-start", "bottom-end", "bottom-center", "bottom-start"])
  attr(:is_hook, :boolean, default: false,
    doc: "Use PureAdminToast JS hook for client-side toast management via push_event")
  attr(:class, :string, default: nil)
  attr(:rest, :global)
  slot(:inner_block)

  def toast_container(assigns) do
    ~H"""
    <div
      id={@id}
      class={build_classes("pa-toast-container", [{"pa-toast-container--#{@position}", true}], @class)}
      phx-hook={if @is_hook, do: "PureAdminToast"}
      data-position={@position}
      {@rest}
    >
      <%= if @inner_block != [], do: render_slot(@inner_block) %>
    </div>
    """
  end

  defp toast_variant_class(assigns) do
    cond do
      assigns.theme_color != nil and assigns.is_filled ->
        "pa-toast--filled-color-#{assigns.theme_color}"
      assigns.theme_color != nil ->
        "pa-toast--color-#{assigns.theme_color}"
      assigns.is_filled ->
        "pa-toast--filled-#{assigns.variant}"
      true ->
        "pa-toast--#{assigns.variant}"
    end
  end
end
